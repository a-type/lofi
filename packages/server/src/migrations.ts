import type { Database } from 'better-sqlite3';
import v0Sql from './migrations/v0.sql.js';
import v1Sql from './migrations/v1.sql.js';
import v2Sql from './migrations/v2.sql.js';
import v3Sql from './migrations/v3.sql.js';
import v4Sql from './migrations/v4.sql.js';

const allMigrations = [v0Sql, v1Sql, v2Sql, v3Sql, v4Sql];

export function migrations(db: Database) {
	// create the versions table if it doesn't exist
	db.prepare(
		`
    CREATE TABLE IF NOT EXISTS versions (
      version INTEGER NOT NULL PRIMARY KEY
    );
  `,
	).run();

	// get the current version
	const versionResult = db
		.prepare(
			`
    SELECT version FROM versions ORDER BY version DESC LIMIT 1;
  `,
		)
		.get() as { version: number } | undefined;
	const version = versionResult?.version ?? 0;

	const sqlToRun = allMigrations.slice(version);

	if (sqlToRun.length === 0) return;

	console.info(
		'Running migrations from ',
		version,
		'to',
		allMigrations.length - 1,
	);

	const run = db.transaction(() => {
		sqlToRun.forEach((sql) => db.exec(sql));
		db.prepare(
			`
			INSERT INTO versions (version) VALUES (?);
		`,
		).run(allMigrations.length);
	});

	run();
}
