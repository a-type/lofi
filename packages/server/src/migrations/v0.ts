export const sql = `
CREATE TABLE IF NOT EXISTS ReplicaInfo (
  id TEXT PRIMARY KEY NOT NULL,
  libraryId TEXT NOT NULL,
  clientId TEXT NOT NULL,
  lastSeenWallClockTime INTEGER,
  ackedLogicalTime TEXT,
  type INTEGER NOT NULL DEFAULT 0
);

CREATE TABLE IF NOT EXISTS OperationHistory (
  libraryId TEXT NOT NULL,
  replicaId TEXT NOT NULL,
  oid TEXT NOT NULL,
  data TEXT NOT NULL,
  timestamp INTEGER NOT NULL,
  PRIMARY KEY (libraryId, replicaId, oid, timestamp)
);

CREATE TABLE IF NOT EXISTS DocumentBaseline (
  oid TEXT PRIMARY KEY NOT NULL,
  snapshot TEXT,
  timestamp TEXT NOT NULL,
  libraryId TEXT NOT NULL
);
`;
