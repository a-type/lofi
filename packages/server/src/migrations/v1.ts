export const sql = `
ALTER TABLE
  ReplicaInfo RENAME TO ReplicaInfoOld;

CREATE TABLE ReplicaInfo (
  id TEXT NOT NULL,
  libraryId TEXT NOT NULL,
  clientId TEXT NOT NULL,
  lastSeenWallClockTime INTEGER,
  ackedLogicalTime TEXT,
  type INTEGER NOT NULL DEFAULT 0,
  PRIMARY KEY (id, libraryId)
);

INSERT INTO
  ReplicaInfo
SELECT
  id,
  libraryId,
  clientId,
  lastSeenWallClockTime,
  ackedLogicalTime,
  type
FROM
  ReplicaInfoOld;

DROP TABLE ReplicaInfoOld;
`;
