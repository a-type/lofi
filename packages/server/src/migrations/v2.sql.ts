// fixes primary key for document baselines to include libraryId

export default `
  ALTER TABLE
    DocumentBaseline RENAME TO DocumentBaselineOld;

  CREATE TABLE DocumentBaseline (
    oid TEXT NOT NULL,
    snapshot TEXT,
    timestamp TEXT NOT NULL,
    libraryId TEXT NOT NULL,
    PRIMARY KEY (libraryId, oid)
  );

  INSERT INTO
    DocumentBaseline
  SELECT
    oid,
    snapshot,
    timestamp,
    libraryId
  FROM
    DocumentBaselineOld;

  DROP TABLE DocumentBaselineOld;
`;
