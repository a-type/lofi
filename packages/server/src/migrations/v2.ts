// fixes primary key for document baselines to include libraryId

export const sql = `
  ALTER TABLE
    DocumentBaseline RENAME TO DocumentBaselineOld;

  CREATE TABLE DocumentBaseline (
    oid TEXT NOT NULL,
    snapshot TEXT,
    timestamp TEXT NOT NULL,
    libraryId TEXT NOT NULL,
    PRIMARY KEY (libraryId, oid)
  );

  INSERT INTO
    DocumentBaseline
  SELECT
    oid,
    snapshot,
    timestamp,
    libraryId
  FROM
    DocumentBaselineOld;

  DROP TABLE DocumentBaselineOld;
`;
