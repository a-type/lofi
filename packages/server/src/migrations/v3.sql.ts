// add the FileMetadata table

export default `
  CREATE TABLE FileMetadata (
    libraryId TEXT NOT NULL,
    fileId TEXT NOT NULL,
    name TEXT NOT NULL,
    type TEXT NOT NULL,
    PRIMARY KEY (libraryId, fileId)
  );
`;
