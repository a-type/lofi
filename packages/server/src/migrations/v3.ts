// add the FileMetadata table

export const sql = `
  CREATE TABLE IF NOT EXISTS FileMetadata (
    libraryId TEXT NOT NULL,
    fileId TEXT NOT NULL,
    name TEXT NOT NULL,
    type TEXT NOT NULL,
    pendingDeleteAt INTEGER,
    PRIMARY KEY (libraryId, fileId)
  );
`;
