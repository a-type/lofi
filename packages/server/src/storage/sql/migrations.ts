import * as v1 from './migrations/v1.js';
import * as v2 from './migrations/v2.js';

export default { v1, v2 };
