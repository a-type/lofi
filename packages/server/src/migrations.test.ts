import { describe, it, expect, afterAll, beforeAll } from 'vitest';
import create from 'better-sqlite3';
import { migrations } from './migrations.js';
import fs from 'fs';

describe('the v5->v6 migration', () => {
	let dbFile: string;
	let db: create.Database;
	beforeAll(() => {
		dbFile = `v5-migration-test-${Math.random()}.sqlite`;
		db = create(dbFile);
	});
	afterAll(() => {
		db.close();
		fs.rmSync(dbFile);
	});

	it('rewrites legacy oids in all ops and baselines', () => {
		// manually migrate to v4 before starting and seed the database with some
		// operations and baselines.
		migrations(db, 5);
		// manually insert some operations and baselines
		db.transaction(() => {
			[
				{
					libraryId: 'test-1',
					replicaId: 'foo',
					oid: 'weirds/b',
					data: {
						op: 'set',
						value: { '@@type': 'ref', id: 'weirds/b.objectMap.one:ghi' },
					},
					timestamp: '1',
					serverOrder: 1,
				},
				{
					libraryId: 'test-1',
					replicaId: 'foo',
					oid: 'weirds/b.list:ghi',
					data: {
						op: 'list-push',
						value: { '@@type': 'ref', id: 'weirds/b.list.#:abc' },
					},
					timestamp: '1',
					serverOrder: 2,
				},
				{
					libraryId: 'test-1',
					replicaId: 'foo',
					oid: 'weirds/b.list:ghi',
					data: {
						op: 'list-remove',
						value: { '@@type': 'ref', id: 'weirds/b.list.#:abc' },
					},
					timestamp: '2',
					serverOrder: 3,
				},
				{
					libraryId: 'test-2',
					replicaId: 'bar',
					oid: 'weirds/b',
					data: {
						op: 'set',
						value: { '@@type': 'ref', id: 'weirds/b.objectMap.one:ghi' },
					},
					timestamp: '1a',
					serverOrder: 1,
				},
			].forEach(
				({ libraryId, replicaId, oid, data, timestamp, serverOrder }) => {
					db.prepare(
						/* sql */ `
        INSERT INTO OperationHistory (libraryId, replicaId, oid, data, timestamp, serverOrder)
        VALUES (?, ?, ?, ?, ?, ?);
      `,
					).run(
						libraryId,
						replicaId,
						oid,
						JSON.stringify(data),
						timestamp,
						serverOrder,
					);
				},
			);
			[
				{
					libraryId: 'test-1',
					oid: 'weirds/b.objectMap.one:ghi',
					snapshot: { '@@type': 'ref', id: 'weirds/b.list.#:abc' },
					timestamp: '1',
				},
				{
					libraryId: 'test-2',
					oid: 'weirds/b.objectMap.one:ghi',
					snapshot: { '@@type': 'ref', id: 'weirds/b.list.#:abc' },
					timestamp: '1a',
				},
			].forEach(({ libraryId, oid, snapshot, timestamp }) => {
				db.prepare(
					/* sql */ `
        INSERT INTO DocumentBaseline (libraryId, oid, snapshot, timestamp)
        VALUES (?, ?, ?, ?);
      `,
				).run(libraryId, oid, JSON.stringify(snapshot), timestamp);
			});
		})();

		migrations(db);

		// check that the operations and baselines have been rewritten
		db.transaction(() => {
			const operations = db
				.prepare(
					/* sql */ `
      SELECT * from OperationHistory;
    `,
				)
				.all();
			expect(operations).toMatchSnapshot();
			const baselines = db
				.prepare(
					/* sql */ `
      SELECT * from DocumentBaseline;
    `,
				)
				.all();
			expect(baselines).toMatchSnapshot();
		})();

		db.close();
	});
});
