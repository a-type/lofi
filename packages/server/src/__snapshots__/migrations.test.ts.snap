// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`the v5->v6 migration > rewrites legacy oids in all ops and baselines 1`] = `
[
  {
    "data": "{"op":"set","value":{"@@type":"ref","id":"weirds/b:ghi"}}",
    "libraryId": "test-1",
    "oid": "weirds/b",
    "replicaId": "foo",
    "serverOrder": 1,
    "timestamp": 1,
  },
  {
    "data": "{"op":"list-push","value":{"@@type":"ref","id":"weirds/b:abc"}}",
    "libraryId": "test-1",
    "oid": "weirds/b:ghi",
    "replicaId": "foo",
    "serverOrder": 2,
    "timestamp": 1,
  },
  {
    "data": "{"op":"list-remove","value":{"@@type":"ref","id":"weirds/b:abc"}}",
    "libraryId": "test-1",
    "oid": "weirds/b:ghi",
    "replicaId": "foo",
    "serverOrder": 3,
    "timestamp": 2,
  },
  {
    "data": "{"op":"set","value":{"@@type":"ref","id":"weirds/b:ghi"}}",
    "libraryId": "test-2",
    "oid": "weirds/b",
    "replicaId": "bar",
    "serverOrder": 1,
    "timestamp": "1a",
  },
]
`;

exports[`the v5->v6 migration > rewrites legacy oids in all ops and baselines 2`] = `
[
  {
    "libraryId": "test-1",
    "oid": "weirds/b:ghi",
    "snapshot": "{"@@type":"ref","id":"weirds/b:abc"}",
    "timestamp": "1",
  },
  {
    "libraryId": "test-2",
    "oid": "weirds/b:ghi",
    "snapshot": "{"@@type":"ref","id":"weirds/b:abc"}",
    "timestamp": "1a",
  },
]
`;
